File: implementations/flask-python/ashurbanipal/db/postgres.py
"""Postgres DbSource implementation (`spec/protocol.md` §5)."""

from __future__ import annotations

import psycopg
from psycopg.types.string import TextLoader

from ..filter import Condition
from . import (
    ColumnInfo,
    ColumnPair,
    ColumnRef,
    DbSource,
    FilterParseError,
    KeyKind,
    NotAllowed,
    QueryOpts,
    ReferencedBy,
    TableData,
    TableInfo,
    group_referenced_by,
    quote_ident,
    wrap_driver_errors,
)

# Bounds catalog queries (`spec/protocol.md` §6).
CATALOG_TIMEOUT_SECS = 5


class _LenientTextLoader(TextLoader):
    """Use the protocol sentinel when psycopg cannot decode text."""

    def load(self, data):
        try:
            return super().load(data)
        except UnicodeDecodeError:
            return "<undecodable>"


def _build_where_clause(conditions: list[Condition], column_names: list[str]) -> tuple[str, list[str]]:
    """Validate filter columns before SQL interpolation (`spec/protocol.md` §5.4.2)."""
    if not conditions:
        return "", []

    values: list[str] = []
    clause_parts: list[str] = []
    for i, cond in enumerate(conditions):
        if cond.column not in column_names:
            raise NotAllowed(f"column {cond.column!r}")
        quoted = quote_ident(cond.column)
        if cond.op in ("IS NULL", "IS NOT NULL"):
            inner = f"{quoted}::text {cond.op}"
        else:
            inner = f"{quoted}::text {cond.op} %s"
            values.append(cond.value)
        wrapped = f"(NOT ({inner}))" if cond.not_ else f"({inner})"

        if i > 0:
            if cond.logic not in ("AND", "OR"):
                raise FilterParseError(f"condition {i} is missing logic")
            clause_parts.append(" AND " if cond.logic == "AND" else " OR ")
        clause_parts.append(wrapped)

    return " where " + "".join(clause_parts), values


class PgSource(DbSource):
    def __init__(self, dsn: str):
        self._dsn = dsn

    def _connect(self) -> psycopg.Connection:
        return psycopg.connect(self._dsn)

    def _list_schemas(self, cur: psycopg.Cursor) -> list[str]:
        cur.execute(
            "select nspname from pg_namespace "
            "where nspname not in ('pg_catalog', 'information_schema') "
            "and nspname not like 'pg_toast%%' "
            "and nspname not like 'pg_temp\\_%%' escape '\\' "
            "and has_schema_privilege(nspname, 'USAGE') "
            "order by nspname"
        )
        return [row[0] for row in cur.fetchall()]

    def _resolve_schema(self, cur: psycopg.Cursor, requested: str | None) -> str:
        schemas = self._list_schemas(cur)
        if requested is not None:
            resolved = requested
        else:
            cur.execute("select current_schema()")
            resolved = cur.fetchone()[0]
        if resolved not in schemas:
            raise NotAllowed(f"schema {resolved!r}")
        return resolved

    def _allowed_tables(self, cur: psycopg.Cursor, schema: str) -> list[str]:
        # Mirrors list_tables' own pg_class/relkind = 'r' predicate, not
        # information_schema.tables' broader BASE TABLE (which also matches
        # partitioned tables) — keeps this allow-list and list_tables in
        # lockstep (docs/adapter-decisions.md §5.2/§5.3).
        cur.execute(
            "select c.relname from pg_class c "
            "join pg_namespace n on n.oid = c.relnamespace "
            "where n.nspname = %s and c.relkind = 'r' "
            "  and has_table_privilege(c.oid, 'SELECT') "
            "order by c.relname",
            (schema,),
        )
        return [row[0] for row in cur.fetchall()]

    def _allowed_columns(self, cur: psycopg.Cursor, schema: str, table: str) -> list[str]:
        cur.execute(
            "select column_name from information_schema.columns "
            "where table_schema = %s and table_name = %s "
            "order by ordinal_position",
            (schema, table),
        )
        return [row[0] for row in cur.fetchall()]

    def _key_metadata(self, cur: psycopg.Cursor, schema: str, table: str) -> tuple[set[str], dict[str, ColumnRef]]:
        """Use constraint_schema for cross-schema FKs; omit composites (`spec/protocol.md` §5.4.1)."""
        cur.execute(
            "select tc.constraint_name, tc.constraint_type, kcu.column_name, "
            "       ccu.table_schema as ref_schema, ccu.table_name as ref_table, "
            "       ccu.column_name as ref_column "
            "from information_schema.table_constraints tc "
            "join information_schema.key_column_usage kcu "
            "  on kcu.constraint_name = tc.constraint_name "
            " and kcu.table_schema = tc.table_schema "
            "left join information_schema.constraint_column_usage ccu "
            "  on ccu.constraint_name = tc.constraint_name "
            " and ccu.constraint_schema = tc.table_schema "
            " and tc.constraint_type = 'FOREIGN KEY' "
            "where tc.table_schema = %s "
            "  and tc.table_name = %s "
            "  and tc.constraint_type in ('PRIMARY KEY', 'FOREIGN KEY')",
            (schema, table),
        )
        pk_columns: set[str] = set()
        fk_candidates: dict[str, list[tuple]] = {}
        for constraint_name, constraint_type, column_name, ref_schema, ref_table, ref_column in cur.fetchall():
            if constraint_type == "PRIMARY KEY":
                pk_columns.add(column_name)
            elif constraint_type == "FOREIGN KEY":
                fk_candidates.setdefault(constraint_name, []).append(
                    (column_name, ref_schema, ref_table, ref_column)
                )

        fk_columns: dict[str, ColumnRef] = {}
        for members in fk_candidates.values():
            distinct_columns = {m[0] for m in members}
            if len(distinct_columns) != 1:
                continue
            column_name, ref_schema, ref_table, ref_column = members[0]
            if ref_schema is None or ref_table is None or ref_column is None:
                continue
            fk_columns[column_name] = ColumnRef(
                table=ref_table,
                column=ref_column,
                schema=ref_schema if ref_schema != schema else None,
            )
        return pk_columns, fk_columns

    @wrap_driver_errors(psycopg.Error)
    def list_schemas(self) -> list[str]:
        with self._connect() as conn, conn.cursor() as cur:
            cur.execute(f"SET LOCAL statement_timeout = '{CATALOG_TIMEOUT_SECS}s'")
            return self._list_schemas(cur)

    @wrap_driver_errors(psycopg.Error)
    def list_tables(self, schema: str | None) -> list[TableInfo]:
        with self._connect() as conn, conn.cursor() as cur:
            cur.execute(f"SET LOCAL statement_timeout = '{CATALOG_TIMEOUT_SECS}s'")
            resolved = self._resolve_schema(cur, schema)
            cur.execute(
                "select c.relname::text, obj_description(c.oid, 'pg_class') "
                "from pg_class c "
                "join pg_namespace n on n.oid = c.relnamespace "
                "where n.nspname = %s and c.relkind = 'r' "
                "  and has_table_privilege(c.oid, 'SELECT') "
                "order by c.relname",
                (resolved,),
            )
            return [TableInfo(name=name, comment=comment) for name, comment in cur.fetchall()]

    @wrap_driver_errors(psycopg.Error)
    def table_counts(self, schema: str | None) -> list[tuple[str, int]]:
        with self._connect() as conn, conn.cursor() as cur:
            cur.execute(f"SET LOCAL statement_timeout = '{CATALOG_TIMEOUT_SECS}s'")
            resolved = self._resolve_schema(cur, schema)
            cur.execute(
                "select c.relname::text, c.reltuples::bigint "
                "from pg_class c "
                "join pg_namespace n on n.oid = c.relnamespace "
                "where n.nspname = %s and c.relkind = 'r' "
                "  and has_table_privilege(c.oid, 'SELECT') "
                "order by c.relname",
                (resolved,),
            )
            return list(cur.fetchall())

    @wrap_driver_errors(psycopg.Error)
    def query_table(self, schema: str | None, table: str, opts: QueryOpts) -> TableData:
        with self._connect() as conn:
            # psycopg snapshots adapters when the cursor is created.
            conn.adapters.register_loader("text", _LenientTextLoader)
            with conn.cursor() as cur:
                cur.execute(f"SET LOCAL statement_timeout = '{int(opts.timeout_secs)}s'")

                resolved_schema = self._resolve_schema(cur, schema)
                tables = self._allowed_tables(cur, resolved_schema)
                if table not in tables:
                    raise NotAllowed(f"table {table!r}")

                column_names = self._allowed_columns(cur, resolved_schema, table)
                sort = None
                if opts.sort is not None:
                    if opts.sort not in column_names:
                        raise NotAllowed(f"column {opts.sort!r}")
                    sort = opts.sort

                where_clause, filter_values = _build_where_clause(opts.filter or [], column_names)

                cur.execute(
                    "select column_name, data_type from information_schema.columns "
                    "where table_schema = %s and table_name = %s "
                    "order by ordinal_position",
                    (resolved_schema, table),
                )
                column_types = list(cur.fetchall())

                # attnum survives dropped columns; ordinal_position does not.
                cur.execute(
                    "select a.attname::text, col_description(a.attrelid, a.attnum::int) "
                    "from pg_attribute a "
                    "join pg_class c on c.oid = a.attrelid "
                    "join pg_namespace n on n.oid = c.relnamespace "
                    "where n.nspname = %s and c.relname = %s "
                    "  and a.attnum > 0 and not a.attisdropped",
                    (resolved_schema, table),
                )
                column_comments = {name: comment for name, comment in cur.fetchall() if comment is not None}

                pk_columns, fk_columns = self._key_metadata(cur, resolved_schema, table)
                columns = [
                    ColumnInfo(
                        name=name,
                        type_name=type_name,
                        key=(KeyKind.PK if name in pk_columns else (KeyKind.FK if name in fk_columns else None)),
                        references=fk_columns.get(name),
                        comment=column_comments.get(name),
                    )
                    for name, type_name in column_types
                ]

                select_list = ", ".join(f"{quote_ident(c.name)}::text" for c in columns)
                order_clause = ""
                if sort is not None:
                    # Qualify the source column to avoid ordering the text-cast output.
                    direction = "desc" if opts.descending else "asc"
                    order_clause = f" order by {quote_ident(table)}.{quote_ident(sort)} {direction}"
                query = (
                    f"select {select_list} from {quote_ident(resolved_schema)}.{quote_ident(table)}"
                    f"{where_clause}{order_clause} limit %s offset %s"
                )
                try:
                    cur.execute(query, (*filter_values, opts.limit, opts.offset))
                except psycopg.errors.InsufficientPrivilege as exc:
                    # The allow-list already rejects tables the role can't
                    # SELECT, so a permission denied here is a residual edge;
                    # report it as NotAllowed (400), not a driver 500.
                    raise NotAllowed(f"table {table!r}") from exc
                pg_rows = cur.fetchall()
                rows = [
                    {
                        col.name: (None if value is None else str(value))
                        for col, value in zip(columns, row, strict=True)
                    }
                    for row in pg_rows
                ]

                cur.execute(
                    "select reltuples::bigint from pg_class c "
                    "join pg_namespace n on n.oid = c.relnamespace "
                    "where n.nspname = %s and c.relname = %s",
                    (resolved_schema, table),
                )
                total_approx = cur.fetchone()[0]

                return TableData(columns=columns, rows=rows, total_approx=total_approx)

    @wrap_driver_errors(psycopg.Error)
    def common_values(self, schema: str | None, table: str, column: str) -> list[tuple[str, float]]:
        with self._connect() as conn, conn.cursor() as cur:
            cur.execute(f"SET LOCAL statement_timeout = '{CATALOG_TIMEOUT_SECS}s'")
            resolved_schema = self._resolve_schema(cur, schema)
            tables = self._allowed_tables(cur, resolved_schema)
            if table not in tables:
                raise NotAllowed(f"table {table!r}")
            columns = self._allowed_columns(cur, resolved_schema, table)
            if column not in columns:
                raise NotAllowed(f"column {column!r}")

            # most_common_vals is anyarray; ::text::text[] reads it
            # uniformly. NULL (no ANALYZE stats yet) unnests to zero rows,
            # not an error.
            cur.execute(
                "select t.val, t.freq "
                "from pg_stats, "
                "     lateral unnest(most_common_vals::text::text[], most_common_freqs) as t(val, freq) "
                "where schemaname = %s and tablename = %s and attname = %s "
                "order by t.freq desc",
                (resolved_schema, table, column),
            )
            rows = list(cur.fetchall())

            cur.execute(
                "select data_type from information_schema.columns "
                "where table_schema = %s and table_name = %s and column_name = %s",
                (resolved_schema, table, column),
            )
            data_type_row = cur.fetchone()
            data_type = data_type_row[0] if data_type_row else None

        # boolean's array-literal text form is "t"/"f", not "true"/"false" —
        # normalize to match query_table's rendering.
        if data_type == "boolean":
            rows = [({"t": "true", "f": "false"}.get(val, val), freq) for val, freq in rows]
        return [(val, float(freq)) for val, freq in rows]

    @wrap_driver_errors(psycopg.Error)
    def referenced_by(self, schema: str | None, table: str) -> list[ReferencedBy]:
        with self._connect() as conn, conn.cursor() as cur:
            cur.execute(f"SET LOCAL statement_timeout = '{CATALOG_TIMEOUT_SECS}s'")
            resolved_schema = self._resolve_schema(cur, schema)

            # Resolved once and bound as %s below (con.confrelid = %s) — the
            # same pg_class/relkind = 'r'/has_table_privilege predicate
            # _allowed_tables uses, so a partitioned or unknown table
            # rejects here instead of a silent [] from the confrelid join
            # finding zero rows (spec/protocol.md §5.2/§5.9).
            cur.execute(
                "select c.oid from pg_class c "
                "join pg_namespace n on n.oid = c.relnamespace "
                "where n.nspname = %s and c.relname = %s and c.relkind = 'r' "
                "  and has_table_privilege(c.oid, 'SELECT')",
                (resolved_schema, table),
            )
            target_oid_row = cur.fetchone()
            if target_oid_row is None:
                raise NotAllowed(f"table {table!r}")
            target_oid = target_oid_row[0]

            # pg_catalog, not information_schema: the reverse (confrelid)
            # filter over the standard-SQL views runs 20-40x slower on a
            # large catalog (docs/adapter-decisions.md §5.9).
            # has_table_privilege is the per-referrer read gate and works
            # across schemas, so cross-schema referrers are reported and
            # gated without a separate allow-list pass. conparentid = 0
            # excludes a partitioned referrer's per-partition constraint
            # copies (Postgres clones the parent's FK onto each partition)
            # — without it, one logical FK fans out into one entry per
            # partition plus the parent.
            cur.execute(
                "select rn.nspname, rc.relname, con.conname, fa.attname, ta.attname "
                "from pg_constraint con "
                "join pg_class rc on rc.oid = con.conrelid "
                "join pg_namespace rn on rn.oid = rc.relnamespace "
                "join lateral unnest(con.conkey, con.confkey) with ordinality "
                "     as k(from_attnum, to_attnum, n) on true "
                "join pg_attribute fa on fa.attrelid = con.conrelid and fa.attnum = k.from_attnum "
                "join pg_attribute ta on ta.attrelid = con.confrelid and ta.attnum = k.to_attnum "
                "where con.contype = 'f' "
                "  and con.conparentid = 0 "
                "  and con.confrelid = %s "
                "  and has_table_privilege(con.conrelid, 'SELECT') "
                "order by rn.nspname, rc.relname, con.conname, k.n",
                (target_oid,),
            )
            rows = list(cur.fetchall())

        return group_referenced_by(
            [
                ReferencedBy(
                    table=ref_table,
                    constraint=constraint,
                    columns=[ColumnPair(from_=from_col, to=to_col)],
                    schema=ref_schema if ref_schema != resolved_schema else None,
                )
                for ref_schema, ref_table, constraint, from_col, to_col in rows
            ]
        )

File: implementations/node-express/src/db/postgres.ts
import type { Pool, PoolClient } from "pg";
import { NotAllowedError, quoteIdent } from "../errors.js";
import { buildWhereClause } from "../filter.js";
import {
  type ColumnInfo,
  type ColumnRef,
  type CommonValueEntry,
  type CountEntry,
  cellToJson,
  type DbSource,
  findExact,
  type QueryOpts,
  type TableData,
  type TableInfo,
} from "./types.js";

/**
 * The default/reference `DbSource` — ported against
 * implementations/rust/src/db/postgres.rs (also cross-checked against
 * implementations/go-nethttp/catalog.go). Every query (catalog/metadata
 * included, not just row fetches) is bounded by the caller-supplied
 * timeoutMs, applied per-connection via `SET LOCAL statement_timeout`
 * inside a transaction so it never leaks onto a pooled connection reused
 * by another request.
 */
export class PostgresSource implements DbSource {
  constructor(private readonly pool: Pool) {}

  // Runs `fn` against a client with statement_timeout bounded for the
  // duration of this one query, inside its own transaction — SET LOCAL is
  // transaction-scoped, so this is the only way to bound a timeout
  // per-query without mutating the pooled connection's session state for
  // whichever request borrows it next.
  private async withTimeout<T>(timeoutMs: number, fn: (client: PoolClient) => Promise<T>): Promise<T> {
    const client = await this.pool.connect();
    try {
      await client.query("BEGIN");
      await client.query(`SET LOCAL statement_timeout = ${timeoutMs}`);
      const result = await fn(client);
      await client.query("COMMIT");
      return result;
    } catch (err) {
      await client.query("ROLLBACK").catch(() => {});
      throw err;
    } finally {
      client.release();
    }
  }

  // Excludes the catalogs themselves (`pg_catalog`, `information_schema`,
  // `pg_toast%`, `pg_temp_%`) and anything the connected role can't
  // actually use, so a schema only ever appears here if it's both a real
  // user namespace and one this role has USAGE on.
  private async allowedSchemas(client: PoolClient): Promise<string[]> {
    const { rows } = await client.query<{ nspname: string }>(
      `select nspname from pg_namespace
       where nspname not in ('pg_catalog', 'information_schema')
         and nspname not like 'pg_toast%'
         and nspname not like 'pg_temp\\_%' escape '\\'
         and has_schema_privilege(nspname, 'USAGE')
       order by nspname`,
    );
    return rows.map((r) => r.nspname);
  }

  // Resolves the schema for one operation exactly once: an explicit
  // request and an absent one (resolved via current_schema()) both go
  // through the same allow-list, so neither path can reach a schema the
  // other would reject (docs/adapter-decisions.md §1). Callers that run
  // more than one query per operation call this once inside their
  // withTimeout transaction, which pins the whole operation to one
  // physical connection — immune to pool sessions with divergent
  // search_path.
  private async resolveSchema(client: PoolClient, requested: string | undefined): Promise<string> {
    const schemas = await this.allowedSchemas(client);
    let resolved: string;
    if (requested !== undefined) {
      resolved = requested;
    } else {
      const { rows } = await client.query<{ current_schema: string }>("select current_schema()");
      resolved = rows[0].current_schema;
    }
    const real = findExact(schemas, resolved);
    if (!real) {
      throw new NotAllowedError(`schema "${resolved}"`);
    }
    return real;
  }

  private async allowedTables(client: PoolClient, schema: string): Promise<string[]> {
    const { rows } = await client.query<{ table_name: string }>(
      `select table_name from information_schema.tables
       where table_schema = $1 and table_type = 'BASE TABLE'
       order by table_name`,
      [schema],
    );
    return rows.map((r) => r.table_name);
  }

  private async allowedColumns(client: PoolClient, schema: string, table: string): Promise<string[]> {
    const { rows } = await client.query<{ column_name: string }>(
      `select column_name from information_schema.columns
       where table_schema = $1 and table_name = $2
       order by ordinal_position`,
      [schema, table],
    );
    return rows.map((r) => r.column_name);
  }

  // Returns the set of primary-key columns and a column -> ColumnRef map
  // for single-column foreign keys. Composite FKs are dropped entirely
  // rather than risk mislabeling which referencing column pairs with which
  // referenced column (spec/protocol.md §5.4.1); composite *primary* keys
  // are NOT dropped this way — every PK column still gets key="pk".
  //
  // The `ccu` join must match on `ccu.constraint_schema` (the schema the
  // constraint itself lives in, always equal to `tc.table_schema`), not
  // `ccu.table_schema` (the schema of the table constraint_column_usage is
  // describing — for a FOREIGN KEY row that's the *referenced* table's
  // schema, which for a cross-schema FK differs from the constraining
  // table's schema). Joining on `ccu.table_schema` instead silently drops
  // every cross-schema FK's metadata (the LEFT JOIN just never matches).
  private async keyMetadata(
    client: PoolClient,
    schema: string,
    table: string,
  ): Promise<{ pkColumns: Set<string>; fkColumns: Map<string, ColumnRef> }> {
    const { rows } = await client.query<{
      constraint_name: string;
      constraint_type: string;
      column_name: string;
      ref_schema: string | null;
      ref_table: string | null;
      ref_column: string | null;
    }>(
      `select tc.constraint_name, tc.constraint_type, kcu.column_name,
              ccu.table_schema as ref_schema, ccu.table_name as ref_table, ccu.column_name as ref_column
       from information_schema.table_constraints tc
       join information_schema.key_column_usage kcu
         on kcu.constraint_name = tc.constraint_name
        and kcu.table_schema = tc.table_schema
       left join information_schema.constraint_column_usage ccu
         on ccu.constraint_name = tc.constraint_name
        and ccu.constraint_schema = tc.table_schema
        and tc.constraint_type = 'FOREIGN KEY'
       where tc.table_schema = $1
         and tc.table_name = $2
         and tc.constraint_type in ('PRIMARY KEY', 'FOREIGN KEY')`,
      [schema, table],
    );

    const pkColumns = new Set<string>();
    const fkCandidates = new Map<
      string,
      { column: string; refSchema: string | null; refTable: string | null; refColumn: string | null }[]
    >();
    for (const row of rows) {
      if (row.constraint_type === "PRIMARY KEY") {
        pkColumns.add(row.column_name);
      } else if (row.constraint_type === "FOREIGN KEY") {
        const list = fkCandidates.get(row.constraint_name) ?? [];
        list.push({
          column: row.column_name,
          refSchema: row.ref_schema,
          refTable: row.ref_table,
          refColumn: row.ref_column,
        });
        fkCandidates.set(row.constraint_name, list);
      }
    }

    const fkColumns = new Map<string, ColumnRef>();
    for (const members of fkCandidates.values()) {
      const distinctColumns = new Set(members.map((m) => m.column));
      if (distinctColumns.size !== 1) continue; // composite FK: omit entirely
      const first = members[0];
      if (first.refSchema && first.refTable && first.refColumn) {
        // Same-schema is the overwhelming common case; omitting `schema`
        // there keeps the wire payload byte-identical to before this field
        // existed.
        const ref: ColumnRef = { table: first.refTable, column: first.refColumn };
        if (first.refSchema !== schema) ref.schema = first.refSchema;
        fkColumns.set(first.column, ref);
      }
    }
    return { pkColumns, fkColumns };
  }

  async listSchemas(timeoutMs: number): Promise<string[]> {
    return this.withTimeout(timeoutMs, (client) => this.allowedSchemas(client));
  }

  async listTables(schema: string | undefined, timeoutMs: number): Promise<TableInfo[]> {
    return this.withTimeout(timeoutMs, async (client) => {
      const realSchema = await this.resolveSchema(client, schema);
      const { rows } = await client.query<{ relname: string; comment: string | null }>(
        `select c.relname::text, obj_description(c.oid, 'pg_class') as comment
         from pg_class c
         join pg_namespace n on n.oid = c.relnamespace
         where n.nspname = $1 and c.relkind = 'r'
         order by c.relname`,
        [realSchema],
      );
      return rows.map((r) => {
        const t: TableInfo = { name: r.relname };
        if (r.comment !== null) t.comment = r.comment;
        return t;
      });
    });
  }

  async tableCounts(schema: string | undefined, timeoutMs: number): Promise<CountEntry[]> {
    return this.withTimeout(timeoutMs, async (client) => {
      const realSchema = await this.resolveSchema(client, schema);
      const { rows } = await client.query<{ relname: string; reltuples: string }>(
        `select c.relname::text, c.reltuples::bigint::text as reltuples
         from pg_class c
         join pg_namespace n on n.oid = c.relnamespace
         where n.nspname = $1 and c.relkind = 'r'
         order by c.relname`,
        [realSchema],
      );
      return rows.map((r) => ({ table: r.relname, approx_rows: Number(r.reltuples) }));
    });
  }

  async queryTable(schema: string | undefined, table: string, opts: QueryOpts, timeoutMs: number): Promise<TableData> {
    return this.withTimeout(timeoutMs, async (client) => {
      const realSchema = await this.resolveSchema(client, schema);
      const tables = await this.allowedTables(client, realSchema);
      const realTable = findExact(tables, table);
      if (!realTable) {
        throw new NotAllowedError(`table "${table}"`);
      }

      const columnNames = await this.allowedColumns(client, realSchema, realTable);

      let sort: string | undefined;
      if (opts.sort !== undefined) {
        sort = findExact(columnNames, opts.sort);
        if (!sort) {
          throw new NotAllowedError(`column "${opts.sort}"`);
        }
      }

      let whereClause = "";
      let filterValues: string[] = [];
      if (opts.filter.length > 0) {
        const built = buildWhereClause(opts.filter, columnNames, 3);
        whereClause = built.where;
        filterValues = built.values;
      }

      const { rows: columnTypeRows } = await client.query<{ column_name: string; data_type: string }>(
        `select column_name, data_type from information_schema.columns
         where table_schema = $1 and table_name = $2
         order by ordinal_position`,
        [realSchema, realTable],
      );

      // Joins through pg_attribute/pg_class directly: col_description is
      // keyed by attnum, which can diverge from ordinal_position once a
      // column has been dropped.
      const { rows: commentRows } = await client.query<{ attname: string; comment: string | null }>(
        `select a.attname::text, col_description(a.attrelid, a.attnum::int) as comment
         from pg_attribute a
         join pg_class c on c.oid = a.attrelid
         join pg_namespace n on n.oid = c.relnamespace
         where n.nspname = $1 and c.relname = $2
           and a.attnum > 0 and not a.attisdropped`,
        [realSchema, realTable],
      );
      const columnComments = new Map<string, string>();
      for (const row of commentRows) {
        if (row.comment !== null) columnComments.set(row.attname, row.comment);
      }

      const { pkColumns, fkColumns } = await this.keyMetadata(client, realSchema, realTable);

      const columns: ColumnInfo[] = columnTypeRows.map((ct) => {
        const col: ColumnInfo = { name: ct.column_name, type: ct.data_type };
        if (pkColumns.has(ct.column_name)) {
          col.key = "pk";
          if (fkColumns.has(ct.column_name)) col.references = fkColumns.get(ct.column_name);
        } else if (fkColumns.has(ct.column_name)) {
          col.key = "fk";
          col.references = fkColumns.get(ct.column_name);
        }
        const comment = columnComments.get(ct.column_name);
        if (comment !== undefined) col.comment = comment;
        return col;
      });

      const selectList = columns.map((col) => `${quoteIdent(col.name)}::text`).join(", ");

      // Table-qualified (by relation name, not schema — a FROM item's
      // correlation name is its own relation name regardless of whether
      // FROM itself is schema-qualified): an unqualified `order by "col"`
      // would resolve to the ::text-cast output column in selectList,
      // sorting lexicographically instead of by the real typed value.
      let orderClause = "";
      if (sort !== undefined) {
        const direction = opts.descending ? "desc" : "asc";
        orderClause = ` order by ${quoteIdent(realTable)}.${quoteIdent(sort)} ${direction}`;
      }

      // Identifiers spliced here are schema-validated (realSchema via
      // resolveSchema, realTable/columns via allowedTables/allowedColumns,
      // sort via the findExact check above, filter columns via
      // buildWhereClause's own allow-list check); every value is a bound
      // $N parameter.
      const query = `select ${selectList} from ${quoteIdent(realSchema)}.${quoteIdent(realTable)}${whereClause}${orderClause} limit $1 offset $2`;
      const args: unknown[] = [opts.limit, opts.offset, ...filterValues];

      const { rows: dataRows } = await client.query(query, args);
      const outRows: Record<string, string | null>[] = dataRows.map((row: Record<string, unknown>) => {
        const out: Record<string, string | null> = {};
        for (const col of columns) {
          out[col.name] = cellToJson(row[col.name]);
        }
        return out;
      });

      const { rows: countRows } = await client.query<{ reltuples: string }>(
        `select reltuples::bigint::text as reltuples from pg_class c
         join pg_namespace n on n.oid = c.relnamespace
         where n.nspname = $1 and c.relname = $2`,
        [realSchema, realTable],
      );
      const totalApprox = countRows.length > 0 ? Number(countRows[0].reltuples) : -1;

      return { columns, rows: outRows, total_approx: totalApprox };
    });
  }

  async commonValues(
    schema: string | undefined,
    table: string,
    column: string,
    timeoutMs: number,
  ): Promise<CommonValueEntry[]> {
    return this.withTimeout(timeoutMs, async (client) => {
      const realSchema = await this.resolveSchema(client, schema);
      const tables = await this.allowedTables(client, realSchema);
      const realTable = findExact(tables, table);
      if (!realTable) {
        throw new NotAllowedError(`table "${table}"`);
      }
      const columnNames = await this.allowedColumns(client, realSchema, realTable);
      const realColumn = findExact(columnNames, column);
      if (!realColumn) {
        throw new NotAllowedError(`column "${column}"`);
      }

      // most_common_vals is anyarray; ::text::text[] reads it uniformly.
      // NULL (no ANALYZE stats yet) unnests to zero rows, not an error.
      const { rows } = await client.query<{ val: string; freq: number }>(
        `select t.val, t.freq
         from pg_stats,
              lateral unnest(most_common_vals::text::text[], most_common_freqs) as t(val, freq)
         where schemaname = $1 and tablename = $2 and attname = $3
         order by t.freq desc`,
        [realSchema, realTable, realColumn],
      );

      const { rows: typeRows } = await client.query<{ data_type: string }>(
        `select data_type from information_schema.columns
         where table_schema = $1 and table_name = $2 and column_name = $3`,
        [realSchema, realTable, realColumn],
      );
      const isBoolean = typeRows.length > 0 && typeRows[0].data_type === "boolean";

      // boolean's array-literal text form is "t"/"f", not "true"/"false" —
      // normalize to match queryTable's rendering.
      return rows.map((r) => {
        let value = r.val;
        if (isBoolean) {
          if (value === "t") value = "true";
          else if (value === "f") value = "false";
        }
        return { value, freq: r.freq };
      });
    });
  }
}
